cdisutils/storage3.py

```diff
@@ -0,0 +1,679 @@
+# -*- coding: utf-8 -*-
+"""
+cdisutils.storage3
+----------------------------------
+
+Utilities for working with object stores using boto3
+
+"""
+from __future__ import print_function
+
+from future import standard_library
+
+standard_library.install_aliases()
+import hashlib
+import json
+import os
+import re
+import sys
+import time
+from builtins import next, object, str, zip
+from io import StringIO as BIO
+from urllib.parse import urlparse
+
+import boto3
+import urllib3
+from botocore.exceptions import ClientError
+
+from .log import get_logger
+
+# NOTE: These are to disable the cert mismatch for our object stores
+# should we ever fix that, we should remove these
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+urllib3.disable_warnings(urllib3.exceptions.SNIMissingWarning)
+
+# magic number here for multipart chunk size, change with care
+DEFAULT_MP_CHUNK_SIZE = 1073741824      # 1GiB
+
+# 16 MiB is used because it was tested for performance, if
+# speed issues are seen, this is a good value to try and
+# tweak. Probably good to keep it powers of 2, and an
+# even interval of the mp_chunk_size above
+DEFAULT_DOWNLOAD_CHUNK_SIZE = 16777216  # 16MiB
+
+
+def get_nearest_file_size(size):
+    """
+    Given a size, in bytes, get the nearest canonical
+    size, and return both the divisor and string size
+    in a tuple
+
+    """
+    sizes = [
+        (1000000000000000, "PB"),
+        (1000000000000, "TB"),
+        (1000000000, "GB"),
+        (1000000, "MB"),
+        (1000, "KB"),
+    ]
+
+    value = sizes[len(sizes) - 1]
+    for entry in sizes:
+        if size < entry[0]:
+            continue
+        value = entry
+        break
+
+    return value
+
+
+def print_running_status(
+    transferred_bytes=None, start_time=None, total_size=None, msg_id=0
+):
+    """ Print the status of a transfer, given time and size """
+    size_info = get_nearest_file_size(transferred_bytes)
+    cur_time = time.clock()
+    base_transfer_rate = float(transferred_bytes) / float(cur_time - start_time)
+    transfer_info = get_nearest_file_size(base_transfer_rate)
+    cur_conv_size = float(transferred_bytes) / float(size_info[0])
+    cur_conv_rate = base_transfer_rate / float(transfer_info[0])
+    if total_size:
+        percent_complete = float(transferred_bytes) / float(total_size) * 100.0
+        sys.stdout.write(
```

Comment:
I'm not sure `sys.stdout.write` will do exactly what we want on Python3, but that can be a problem for when we update the code consuming it to Python 3.

Reply:
We could use a progress bar instead if this becomes an issue.